File: same_tree.cpp
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if(!p && !q) return true;
        if(!p || !q) return false;
        if(p -> val != q -> val) return false;

        int left = isSameTree(p -> left, q -> left);
        int right = isSameTree(p -> right, q -> right);

        return left && right;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(n)

File: same_tree.go
func isSameTree(p *TreeNode, q *TreeNode) bool {
    if(p == nil && q == nil) {
        return true
    } 
    if(p == nil || q == nil) {
        return false
    } 
    if(p.Val != q.Val) {
        return false
    }

    left, right := isSameTree(p.Left, q.Left), isSameTree(p.Right, q.Right)
    return left && right
}

// Time Complexity : O(n)
// Space Complexity : O(n)